pkg/tenant/topics.go

```diff
@@ -237,24 +254,100 @@ func (s *Server) TopicUpdate(c *gin.Context) {
 		return
 	}
 
+	// Sanity check that the ID in the URL matches the ID in the request body.
+	if topic.ID != topicID.String() {
+		log.Warn().Msg("topic id in request body does not match topic id in URL")
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("id in request body does not match id in URL"))
+		return
+	}
+
 	// Verify the topic name exists and return a 400 response if it doesn't.
 	if topic.Name == "" {
 		c.JSON(http.StatusBadRequest, api.ErrorResponse("topic name is required"))
 		return
 	}
 
-	// Get the specified topic from the database and return a 404 response if
-	// it cannot be retrieved.
+	// Topic state is optional but can only be set to readonly.
+	if topic.State != "" && topic.State != pb.TopicTombstone_READONLY.String() {
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("topic state can only be set to READONLY"))
+		return
+	}
+
+	// Fetch the topic metadata from the database.
 	var t *db.Topic
-	if t, err = db.RetrieveTopic(c.Request.Context(), topicID); err != nil {
+	if t, err = db.RetrieveTopic(ctx, topicID); err != nil {
+		if errors.Is(err, db.ErrNotFound) {
+			log.Warn().Err(err).Str("topicID", topicID.String()).Msg("topic not found")
+			c.JSON(http.StatusNotFound, api.ErrorResponse("topic not found"))
+			return
+		}
 		log.Error().Err(err).Str("topicID", topicID.String()).Msg("could not retrieve topic")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not update topic"))
+		return
+	}
+
+	// Ensure the new name is valid
+	t.Name = topic.Name
```

Comment:
The name will only be updated in Tenant, it won't be updated in Ensign. That's probably fine for now but we should create a follow on story for updating the name in Tenant.